underscore.comparison.islike.js

```diff
@@ -0,0 +1,70 @@
+/*
+* Tests if an object is like another. This means objects should follow the same
+* structure and arrays should contain the same types.
+*
+* E.g.
+*
+* _.islike(
+*   {name: "James", age: 10, hobbies: ["football", "computer games", "baking"]},
+*   {name: "", age: 0, hobbies: [""]}
+* )
+*/
+(function() {
+  // Establish the root object, `window` in the browser, or `require` it on the server.
+  if (typeof exports === 'object') {
+    _ = module.exports = require('underscore');
+  }
+
+    var islike = function(obj, pattern) {
+        if (typeof pattern === "function") {
+            return obj instanceof pattern;
+        }
+
+        if (typeof obj !== typeof pattern) return false;
+            if (pattern instanceof Array && !(obj instanceof Array)) return false;
+
+                var type = typeof pattern;
+
+                if (type == "object") {
+                    if (pattern instanceof Array) {
+                        if (pattern.length > 0) {
+                            var oTypes = _.uniq(_.map(obj, fTypeof));
+                            var pTypes = _.uniq(_.map(pattern, fTypeof));
+                            if (_.difference(oTypes, pTypes).length) {
+                                return false;
+                            }
+                        }
+                    } else { // object
+                        if (pattern.constructor === pattern.constructor.prototype.constructor) {
+                            // for 'simple' objects we enumerate
+                            for (var k in pattern) {
+                                if (!pattern.hasOwnProperty(k)) {
```

Comment:
Like `_.any(pattern, function(p, k) { return !islike(obj[k], p); })` then?

I'm not sure if I prefer the for loop, it's slightly longer but a bit clearer and maybe quicker too since we don't need to create a lot of little closures.

Also, since `_.any` uses `_.keys` which iterates the whole object we do lose a bit of the short circuiting benefit of the loop.